python/requirements.txt

```diff
@@ -4,112 +4,65 @@
 #
 # In short, if you change it here, PLEASE also change it in setup.py.
 
+# You can obtain this list from the ray.egg-info/requires.txt
```

Comment:
Thinking about this some more, do we actually need this file if we already have `ray[all]`?

In the `Dockerfile` we can do `pip install -U ray[all] -r requirements_tune.txt -r requirements_train.txt ...`

Reply:
The main reason is actually that we install these requirements before we install ray. Thus if we used `ray[all]` here, we would download the latest ray release, install it, then later compile ray. We've ran into problems with that before...

I'm also not sure if the legacy pip resolver will obey the constraints in `ray[all]` in the same way it obeys constraints in `-c requirements.txt`. The legacy resolver installs all packages as defined on their first level (so `ray` and everything in the provided files), but for second level dependencies (and I think `ray[all]`'s dependencies would be that) it installs the latest versions and accepts incompatibilities.

We want to make sure that Ray's constraints are closely followed. Thus I would like to keep it for now

Reply:
Ok makes sense!

Do you have any suggestions on how to make sure `setup.py` doesn't diverge from `requirements.txt` in the future?

Reply:
I stumbled upon this PR because I saw `--use-deprecated=legacy-resolver` in the Dockerfile above. This will allow for conflicting additions/changes in those downstream files to go through unnoticed in CI, and I would recommend against it.

Regarding the diverging @amogkam, and the addition of `-c/--constraint` across the place in this PR: I would also recommend against it. In general, python libraries (not applications) should have compatibility ranges in their `install_requires` and `extras_require`, and only use exact pins if absolutely necessary due to a specific reason best documented inline.

That said, a lockfile containing a known working set of exhaustive dependencies is nice to have checked into git, and probably even necessary to keep CI reproducible. At work, we generate them with [pipgrip](https://github.com/ddelange/pipgrip).

We specify compatibility ranges for all our deps whenever possible:
- `requests~=2.11` which is equivalent to `requests>=2.11,<3`
- `requests~=2.11.1` which is equivalent to `requests>=2.11.1,<2.12`
to allow interoperability between our libs.

For local development and CI of a lib, we also have a file checked in with the latest known working set. If any txt file in the `./requirements` folder changes in a PR, running a refresh of the lockfile is required by CI. This has numerous advantages, and in practice avoids so much dephell and wasted dev hours, because finding the downstream culprit becomes more or less trivial.

For the devs we have `make lockfile` and `make install`:
```Makefile
.PHONY: install
## Install this repo, plus dev requirements, in editable mode
install:
	pip install -r requirements/ci.txt -r requirements/docs.txt -e .[all] --constraint=requirements/pipgrip.lock

.PHONY: lockfile
## Create a lockfile for all python requirements.
lockfile:
	pipgrip -vv --sort --lock --tree .[all]
	mv pipgrip.lock requirements/pipgrip.lock
```

All our `[extras]` have their own file in that folder, and we have ci.txt and docs.txt separate (not specified in setup.py), because they are not needed by an end user doing `pip install .[all]`.

I think a similar setup would save ray maintainers a lot of future headache. 

Hope that helps shed some light on dephell for big libraries with many extras :)

Reply:
So we only use `--constraint` in `make install`, which is the command used for:
- local development
- CI
- production Docker image